File: backend/src/main/java/br/ufpr/sept/so2/shared/api/ProblemResponses.java
package br.ufpr.sept.so2.shared.api;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.net.URI;
import java.time.OffsetDateTime;
import java.util.LinkedHashMap;
import java.util.Map;

@Component
public class ProblemResponses {

    public static final String ERROR_BASE = "https://secretariaonline.ufpr.br/errors/";

    private final ObjectMapper objectMapper;

    public ProblemResponses(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public ProblemDetail create(HttpStatus status, String title, String detail, String type) {
        ProblemDetail problem = ProblemDetail.forStatusAndDetail(status, detail);
        problem.setTitle(title);
        problem.setType(URI.create(ERROR_BASE + type));
        problem.setProperty("timestamp", OffsetDateTime.now().toString());
        return problem;
    }

    public void write(
            HttpServletRequest request,
            HttpServletResponse response,
            HttpStatus status,
            String title,
            String detail,
            String type
    ) throws IOException {
        write(request, response, status, title, detail, type, Map.of());
    }

    public void write(
            HttpServletRequest request,
            HttpServletResponse response,
            HttpStatus status,
            String title,
            String detail,
            String type,
            Map<String, Object> extras
    ) throws IOException {
        if (response.isCommitted()) {
            return;
        }
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("type", ERROR_BASE + type);
        body.put("title", title);
        body.put("status", status.value());
        body.put("detail", detail);
        body.put("instance", request.getRequestURI());
        body.put("timestamp", OffsetDateTime.now().toString());
        extras.forEach(body::put);
        response.setStatus(status.value());
        response.setContentType("application/problem+json");
        objectMapper.writeValue(response.getWriter(), body);
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/shared/api/ProblemResponses.kt
package br.ufpr.sept.so2.shared.api

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.stereotype.Component
import java.net.URI
import java.time.OffsetDateTime

@Component
class ProblemResponses(
    private val objectMapper: ObjectMapper,
) {
    fun create(status: HttpStatus, title: String, detail: String, type: String): ProblemDetail {
        val problem = ProblemDetail.forStatusAndDetail(status, detail)
        problem.title = title
        problem.type = URI.create(ERROR_BASE + type)
        problem.setProperty("timestamp", OffsetDateTime.now().toString())
        return problem
    }

    @JvmOverloads
    fun write(
        request: HttpServletRequest,
        response: HttpServletResponse,
        status: HttpStatus,
        title: String,
        detail: String,
        type: String,
        extras: Map<String, Any> = emptyMap(),
    ) {
        if (response.isCommitted) {
            return
        }
        val body = linkedMapOf<String, Any>(
            "type" to (ERROR_BASE + type),
            "title" to title,
            "status" to status.value(),
            "detail" to detail,
            "instance" to request.requestURI,
            "timestamp" to OffsetDateTime.now().toString(),
        )
        extras.forEach { (key, value) -> body[key] = value }
        response.status = status.value()
        response.contentType = "application/problem+json"
        objectMapper.writeValue(response.writer, body)
    }

    companion object {
        const val ERROR_BASE: String = "https://secretariaonline.ufpr.br/errors/"
    }
}
